lib/portlayer/attach/handle.go

```diff
@@ -0,0 +1,179 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package attach
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/vic/lib/portlayer/exec"
+	"github.com/vmware/vic/pkg/trace"
+	"golang.org/x/net/context"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+const (
+	serialOverLANPort = 8080
+)
+
+// Handler defines the interface
+type Handler interface {
+	Join() error
+	Bind() error
+	Unbind() error
+}
+
+// SerialHandler implements the Handler interface
+type SerialHandler struct {
+	handle *exec.Handle
+}
+
+// NewSerialHandler returns a new SerialHandler
+func NewSerialHandler(handle *exec.Handle) Handler {
+	defer trace.End(trace.Begin(""))
+
+	// make sure a spec exists
+	handle.SetSpec(nil)
+
+	return &SerialHandler{
+		handle: handle,
+	}
+}
+
+func (s *SerialHandler) toggle(connected bool) error {
+	ctx := context.Background()
+
+	// get the virtual device list
+	devices, err := s.handle.Container.Devices(ctx)
+	if err != nil {
+		return err
+	}
+
+	// select the virtual serial ports
+	serials := devices.SelectByBackingInfo((*types.VirtualSerialPortURIBackingInfo)(nil))
+	if len(serials) == 0 {
+		return fmt.Errorf("Unable to bind a device with desired backing")
+	}
+	if len(serials) > 1 {
+		return fmt.Errorf("Multiple matches found with desired backing")
+	}
+	serial := serials[0]
+	log.Debugf("Found a device with desired backing: %#v", serial)
+
+	if serial.GetVirtualDevice().Connectable.Connected == connected {
+		log.Debugf("Already in the desired state (connected: %t)", connected)
+		return nil
+	}
+
+	log.Debugf("Setting the connected state to %t", connected)
+	// set the values
+	serial.GetVirtualDevice().Connectable.Connected = connected
+	if connected && s.handle.ExecConfig.Sessions[s.handle.ExecConfig.ID].Attach {
+		log.Debugf("Setting the start connected state to %t", connected)
+		serial.GetVirtualDevice().Connectable.StartConnected = s.handle.ExecConfig.Sessions[s.handle.ExecConfig.ID].Attach
+	}
+
+	config := &types.VirtualDeviceConfigSpec{
+		Device:    serial,
+		Operation: types.VirtualDeviceConfigSpecOperationEdit,
+	}
+	s.handle.Spec.DeviceChange = append(s.handle.Spec.DeviceChange, config)
+
+	return nil
+}
+
+// Join adds 1 network backed, 2 file backed serial port to the caller and configures them
+func (s *SerialHandler) Join() error {
+	defer trace.End(trace.Begin(""))
+
+	// FIXME: THERE MUST BE ANOTHER WAY
+	// following is from Create@exec.go
+	ips, err := net.LookupIP(exec.ManagementHostName)
```

Comment:
We need to have the serial port configured during Join, however we may need the IP configured during Bind. What occurs after appliance restart and/or if the DHCP address changes?